File: workflow-sdk-java/src/main/java/com/batnam/rochallor_engine/workflow_sdk_java/metrics/Metrics.java
package com.batnam.rochallor_engine.workflow_sdk_java.metrics;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;

/**
 * Prometheus-compatible metrics for the Java SDK, backed by Micrometer.
 */
public class Metrics {

    public final Timer   pollLatency;
    public final Counter lockConflicts;
    public final Timer   handlerLatency;
    public final Counter retriesTotal;
    public final Counter jobsCompleted;
    public final Counter jobsFailed;

    public Metrics(MeterRegistry registry) {
        this.pollLatency = Timer.builder("workflow_sdk_poll_latency_seconds")
                .description("Latency of PollJobs calls")
                .register(registry);

        this.lockConflicts = Counter.builder("workflow_sdk_lock_conflicts_total")
                .description("Poll rounds that returned zero jobs")
                .register(registry);

        this.handlerLatency = Timer.builder("workflow_sdk_handler_latency_seconds")
                .description("Duration of handler executions")
                .register(registry);

        this.retriesTotal = Counter.builder("workflow_sdk_retries_total")
                .description("Jobs retried by the SDK runner")
                .register(registry);

        this.jobsCompleted = Counter.builder("workflow_sdk_jobs_completed_total")
                .tag("outcome", "success")
                .description("Jobs successfully completed")
                .register(registry);

        this.jobsFailed = Counter.builder("workflow_sdk_jobs_completed_total")
                .tag("outcome", "failure")
                .description("Jobs failed")
                .register(registry);
    }
}

File: workflow-sdk-go/metrics/metrics.go
// Package metrics registers Prometheus metrics for the workflow Go SDK.
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	// PollLatency is a histogram of the time from poll call to first job lock.
	PollLatency = promauto.NewHistogram(prometheus.HistogramOpts{
		Name:    "workflow_sdk_poll_latency_seconds",
		Help:    "Latency of PollJobs calls in seconds.",
		Buckets: prometheus.DefBuckets,
	})

	// LockConflicts counts poll rounds that returned zero jobs (empty queue or contention).
	LockConflicts = promauto.NewCounter(prometheus.CounterOpts{
		Name: "workflow_sdk_lock_conflicts_total",
		Help: "Number of poll rounds that returned zero jobs.",
	})

	// HandlerLatency is a histogram of handler execution time, labelled by jobType.
	HandlerLatency = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "workflow_sdk_handler_latency_seconds",
		Help:    "Duration of handler executions in seconds.",
		Buckets: prometheus.DefBuckets,
	}, []string{"job_type"})

	// RetriesTotal counts retry enqueues from the SDK side.
	RetriesTotal = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "workflow_sdk_retries_total",
		Help: "Number of jobs retried by the SDK runner.",
	}, []string{"job_type"})

	// JobsCompletedTotal counts completed jobs, labelled by outcome.
	JobsCompletedTotal = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "workflow_sdk_jobs_completed_total",
		Help: "Number of jobs completed, labelled by outcome: success or failure.",
	}, []string{"job_type", "outcome"})
)
